pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -184,3 +185,68 @@ def add_int_cut(var_values, solve_data, config, feasible=False):
     #     MindtPy.MindtPy_linear_cuts.integer_cuts.add(expr=int_cut)
     # else:
     #     MindtPy.MindtPy_linear_cuts.feasible_integer_cuts.add(expr=int_cut)
+
+
+def add_affine_cuts(solve_data, config):
+    m = solve_data.mip
+    config.logger.info("Adding affine cuts.")
+    counter = 0
+
+    for constr in m.MindtPy_utils.constraint_list:
+        if constr.body.polynomial_degree() in (1, 0):
+            continue
+
+        vars_in_constr = list(
+            identify_variables(constr.body))
+        if any(var.value is None for var in vars_in_constr):
+            continue  # a variable has no values
+
+        # mcpp stuff
+        try:
+            mc_eqn = mc(constr.body)
+        except MCPP_Error as e:
+            config.logger.debug(
+                "Skipping constraint %s due to MCPP error %s" % (constr.name, str(e)))
+            continue  # skip to the next constraint
+        # TODO: check if the value of ccSlope and cvSlope is not Nan or inf. If so, we skip this.
+        ccSlope = mc_eqn.subcc()
+        cvSlope = mc_eqn.subcv()
+        ccStart = mc_eqn.concave()
+        cvStart = mc_eqn.convex()
+
+        valid = True
+        for var in vars_in_constr:
+            if not var.fixed:
+                if ccSlope[var] == float('nan') or ccSlope[var] == float('inf') or cvSlope[var] == float('nan') or cvSlope[var] == float('inf'):
+                    valid = False
```

Comment:
This is not necessarily true. The convex slope (`cvSlope`) can be undefined but we can still add a valid cut for the concave part. I would split this conditional, with its corresponding cut.

Reply:
Quick question - how are cuts added for undefined slopes?

Reply:
Good question. The answer (and something Zedong worked out) is that we should not add cuts in those cases.